keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -0,0 +1,337 @@
+
+- Stop re-configuring the load balancers' node set for fully deleted /
+  newly added cluster nodes, or for nodes which get annotated with
+  `node.kubernetes.io/exclude-from-external-load-balancers`.
+
+## Proposal
+
+### Risks and Mitigations
+
+#### Risk
+
+- Cloud providers which do not allow VM deletion when the VM is referenced by
+  other constructs, will block the cluster auto-scaler (CA) from deleting the VM
+  upon downscale. This will result in reduced downscale performance by the CA,
+  or completely block VM deletion from happening - this is because the service
+  controller will never proceed to de-reference the VM from the load balancer
+  node set until the Node is fully deleted in the API server, which will never
+  occur until the VM is deleted. The three major cloud providers (GCP/AWS/Azure)
+  do however support this, and it is not expected that other providers don't.
+- By removing the b) predicate above we are delaying the removal of the node
+  from the load balancers' node set until the node is completely deleted in the
+  API server. Workloads relying on this behavior will observe a change in
+  behavior, but this also not expected.
+
+#### Mitigations
+
+- Cloud providers/workloads which do not support the behavior mentioned in
+  (Risk)[#risk], have the possibility to set the feature flag to false, thus
+  default back to the current mechanism.
+
+## Design Details
+
+- Implement the change in the service controller and ensure it does not add /
+  remove nodes from the load balancers' node set for cases a) and b) mentioned
+  in (Summary)[#Summary]
+- Add the feature gate: `LoadBalancerConnectionDraining`, set it to "on" by
+  default and promote it directly to GA
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+The service controller in the KCCM currently has a set of tests validating
+expected syncs caused by Node predicates, these will need to be updated.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+Kubernetes in general needs to extended its load balancing test suite with
+disruption tests, this might be the right effort we need to get that ball
+rolling. Testing would include:
+
+- validation that an application running on a deleting VM benefits from graceful
+  termination of its TCP connection.
+- validation that Node readiness state changes to not result in load balancer
+  re-syncs.
+
+### Graduation Criteria
+
+#### GA
+
+This is addressing a sub-optimal solution currently existing in Kubernetes, so
+the feature gate will be GA and "on" by default from the start.
+
+The feature flag should be kept available until we get sufficient evidence of
+people not being affected by this - by not being able to delete VMs on
+downscaling by the cluster auto-scaler.
+
+Tentative timeline is to lock the feature flag to true in 1.27.
+
+### Upgrade / Downgrade Strategy
+
+N/A
```

Comment:
Can you just mention what happens if, for example, someone upgrades to 1.27, gets this feature, then downgrades while they have unready nodes?  I guess it would just remove them all in the next sync loop, right?

Reply:
> I guess it would just remove them all in the next sync loop, right?

Yes, that's what I forsee too. I will mention it

As a question to that though: does Kubernetes even support control plane downgrades? What is supported in those cases?

Reply:
Yes we do, though it is not as common as one might hope.  Every KEP has to think about this case - what happens if we rollback - and for all API changes this mandates an alpha stage.